File: gift1/gift1.c
/*
ID: zeng.aa1
LANG: C
TASK: gift1
*/

#include <stdio.h>
#include <string.h>

int find (char name[256], char names[][256], int NP) {
    int i;
    for (i = 0; i < NP; i++)
        if (!strcmp(name, names[i]))
            return i;
    return -1;
}

int main (void) {
    FILE *f, *g;
    f = fopen ("gift1.in", "r");
    g = fopen ("gift1.out", "w");
    int NP;
    fscanf (f, "%d", &NP);
    char names[NP][256], name[256];
    int money[NP], i, j, k, u, m, d;
    for (i = 0; i < NP; i++) {
        fscanf (f, "%s", names[i]);
        money[i] = 0;
    }
    for (i = 0; i < NP; i++) {
        fscanf (f, "%s", name);
        j = find (name, names, NP);
        fscanf (f, "%d %d", &m, &d);
        for (k = 0; k < d; k++) {
            fscanf (f, "%s", name);
            u = find (name, names, NP);
            money[j] -= m / d;
            money[u] += m / d;
        }
    }
    for (i = 0; i < NP; i++)
        fprintf (g, "%s %d\n", names[i], money[i]);
    fclose (f); fclose (g);
    return 0;
}

File: gift1/gift1.java
/*
ID: zeng.aa1
LANG: JAVA
TASK: gift1
 */

import java.util.*;
import java.io.*;

public class gift1
{
    public static int find( String name, String[] names )
    {
        for ( int i = 0; i < names.length; i++ )
            if ( name.equals( names[i] ) )
                return i;
        
        return -1;
    }

    public static void main( String[] args ) throws IOException
    {
        Scanner in = new Scanner( new File( "gift1.in") );
        FileWriter fw = new FileWriter( "gift1.out" );
        
        int NP = in.nextInt();
        in.nextLine();
        
        String[] names = new String[NP];
        String name;
        int[] money = new int[NP];
        
        for ( int i = 0; i < NP; i++ )
            names[i] = in.nextLine();
        
        for ( int i = 0; i < NP; i++ )
        {
            name = in.nextLine();
            int j = find( name, names );
            StringTokenizer st = new StringTokenizer( in.nextLine() );
            int m = Integer.parseInt( st.nextToken() );
            int d = Integer.parseInt( st.nextToken() );
            for ( int k = 0; k < d; k++ )
            {
                name = in.nextLine();
                int u = find( name, names );
                money[j] -= m / d;
                money[u] += m / d;
            }
        }
        
        for ( int i = 0; i < NP; i++ )
            fw.write( String.format( "%s %d\n", names[i], money[i] ) );
        
        in.close();
        fw.close();
    }
}
